src/sympc/tensor/replicatedshare_tensor.py

```diff
@@ -306,63 +307,104 @@ def rsub(
         """
         return self.__apply_op(y, "sub")
 
-    def mul(self, y):
+    def mul(self, y: Union[int, float, torch.Tensor]) -> "ReplicatedSharedTensor":
         """Apply the "mul" operation between "self" and "y".
 
         Args:
             y: self*y
 
+        Returns:
+            ReplicatedSharedTensor: Result of the operation.
+
         """
+        y, session_vals = self.sanity_checks(self, y)
+
+        is_private = isinstance(y, ReplicatedSharedTensor)
+
+        if is_private:
+
+            from sympc.protocol import Falcon
+
+            result = [Falcon.multiplication_protocol(self, y)]
+
+        else:
+
+            result = [share * y.shares[0] for share in self.shares]
+
+        tensor = ReplicatedSharedTensor(
+            ring_size=self.ring_size, session_uuid=self.session_uuid, config=self.config
+        )
+        tensor.shares = result
```

Comment:
I think here we could modify , if we incorporate the falcon change as only `return self.__apply_op(y,"mul")`

Reply:
Apply_op becomes crowded with lots of if-else statements. sharetensor it makes sense because they outsource it to pytorch.